File: swift/Sources/MDoc/DeviceResponse.swift
import CborCose

/// One document inside a `DeviceResponse` (ISO 18013-5 §8.3.2.1.2.2), reader-side view.
public struct ResponseDocument {
    public let docType: String
    public let issuerSigned: IssuerSigned
    /// `DeviceNameSpacesBytes` (#6.24) as received — for `DeviceAuthentication` reconstruction.
    public let deviceNameSpacesBytes: Cbor
    /// COSE_Sign1 device authentication, or nil when the document uses `deviceMac` instead.
    public let deviceSignature: CoseSign1?
    /// COSE_Mac0 device authentication (ISO 18013-5 §9.1.3.5), for key-agreement device keys.
    public let deviceMac: CoseMac0?
}

/// A wallet's `DeviceResponse` as parsed by the reader/verifier.
public struct DeviceResponse {
    public let version: String
    public let status: Int64
    public let documents: [ResponseDocument]

    public static func decode(_ bytes: [UInt8]) throws -> DeviceResponse {
        guard case let .map(entries) = try CborDecoder.decode(bytes) else { throw MdocError("DeviceResponse must be a map") }
        guard case let .text(version)? = mdocField(entries, "version") else { throw MdocError("missing version") }
        var status: Int64 = 0
        if case let .uint(s)? = mdocField(entries, "status") { status = Int64(s) }

        var documents: [ResponseDocument] = []
        if case let .array(docs)? = mdocField(entries, "documents") {
            documents = try docs.map { doc in
                guard case let .map(docMap) = doc else { throw MdocError("document must be a map") }
                guard case let .text(docType)? = mdocField(docMap, "docType") else { throw MdocError("missing docType") }
                guard let issuerSignedCbor = mdocField(docMap, "issuerSigned") else { throw MdocError("missing issuerSigned") }
                let issuerSigned = try IssuerSigned.fromCbor(issuerSignedCbor)
                guard case let .map(deviceSigned)? = mdocField(docMap, "deviceSigned") else { throw MdocError("missing deviceSigned") }
                guard let deviceNameSpacesBytes = mdocField(deviceSigned, "nameSpaces") else { throw MdocError("missing deviceSigned nameSpaces") }
                guard case let .map(deviceAuth)? = mdocField(deviceSigned, "deviceAuth") else { throw MdocError("missing deviceAuth") }
                let deviceSignature = try mdocField(deviceAuth, "deviceSignature").map { try CoseSign1.fromCbor($0) }
                let deviceMac = try mdocField(deviceAuth, "deviceMac").map { try CoseMac0.fromCbor($0) }
                if deviceSignature == nil, deviceMac == nil { throw MdocError("deviceAuth has neither deviceSignature nor deviceMac") }
                return ResponseDocument(docType: docType, issuerSigned: issuerSigned, deviceNameSpacesBytes: deviceNameSpacesBytes,
                                        deviceSignature: deviceSignature, deviceMac: deviceMac)
            }
        }
        return DeviceResponse(version: version, status: status, documents: documents)
    }
}

File: kotlin/mdoc/src/main/kotlin/com/hopae/eudi/wallet/mdoc/DeviceResponse.kt
package com.hopae.eudi.wallet.mdoc

import com.hopae.eudi.wallet.cbor.Cbor
import com.hopae.eudi.wallet.cbor.CborDecoder
import com.hopae.eudi.wallet.cbor.cose.CoseMac0
import com.hopae.eudi.wallet.cbor.cose.CoseSign1

/** One document inside a `DeviceResponse` (ISO 18013-5 §8.3.2.1.2.2), reader-side view. */
class ResponseDocument(
    val docType: String,
    val issuerSigned: IssuerSigned,
    /** `DeviceNameSpacesBytes` (#6.24) as received — for `DeviceAuthentication` reconstruction. */
    val deviceNameSpacesBytes: Cbor,
    /** COSE_Sign1 device authentication, or null when the document uses [deviceMac] instead. */
    val deviceSignature: CoseSign1?,
    /** COSE_Mac0 device authentication (ISO 18013-5 §9.1.3.5), for key-agreement device keys. */
    val deviceMac: CoseMac0? = null,
)

/** A wallet's `DeviceResponse` as parsed by the reader/verifier. */
class DeviceResponse(val version: String, val status: Long, val documents: List<ResponseDocument>) {

    companion object {
        fun decode(bytes: ByteArray): DeviceResponse {
            val map = CborDecoder.decode(bytes) as? Cbor.CborMap ?: throw MdocException("DeviceResponse must be a map")
            val version = (map.field("version") as? Cbor.Text)?.value ?: throw MdocException("missing version")
            val status = (map.field("status") as? Cbor.UInt)?.value?.toLong() ?: 0L
            val documents = (map.field("documents") as? Cbor.Array)?.items?.map { doc ->
                val docMap = doc as? Cbor.CborMap ?: throw MdocException("document must be a map")
                val docType = (docMap.field("docType") as? Cbor.Text)?.value ?: throw MdocException("missing docType")
                val issuerSigned = IssuerSigned.fromCbor(docMap.field("issuerSigned") ?: throw MdocException("missing issuerSigned"))
                val deviceSigned = docMap.field("deviceSigned") as? Cbor.CborMap ?: throw MdocException("missing deviceSigned")
                val deviceNameSpacesBytes = deviceSigned.field("nameSpaces") ?: throw MdocException("missing deviceSigned nameSpaces")
                val deviceAuth = deviceSigned.field("deviceAuth") as? Cbor.CborMap ?: throw MdocException("missing deviceAuth")
                val deviceSignature = deviceAuth.field("deviceSignature")?.let { CoseSign1.fromCbor(it) }
                val deviceMac = deviceAuth.field("deviceMac")?.let { CoseMac0.fromCbor(it) }
                if (deviceSignature == null && deviceMac == null) throw MdocException("deviceAuth has neither deviceSignature nor deviceMac")
                ResponseDocument(docType, issuerSigned, deviceNameSpacesBytes, deviceSignature, deviceMac)
            } ?: emptyList()
            return DeviceResponse(version, status, documents)
        }

        private fun Cbor.CborMap.field(name: String): Cbor? =
            entries.firstOrNull { (k, _) -> (k as? Cbor.Text)?.value == name }?.second
    }
}
